### ISSUE TITLE
Is the final-sigmoid layer used only for the purpose of viewing the output?

### ISSUE THREAD

**Issue by StuckinPhD** [wolny/pytorch-3dunet#17]
Hi,

This is not an issue with the code but I wanted to understand something.

Based on the comment you made in ``unet3d\model.py``:

```
# apply final_activation (i.e. Sigmoid or Softmax) only for prediction. During training the network outputs
        # logits and it's up to the user to normalize it before visualising with tensorboard or computing accuracy
        if not self.training:
            x = self.final_activation(x)
```

So, when using dice, the final sigmoid layer is only used so that we can view the network output? Is that its only function? If so then why do we give the ``final-sigmoid`` parameter during training as mentioned here:

> In case of DiceLoss and GeneralizedDiceLoss the final score is the average across channels. final_sigmoid=True has to be present in the config when training the network with any of the above losses (and similarly final_sigmoid=True has to be passed to the predict.py if the network was trained with final_sigmoid=True)

Would the network train similarly if the final sigmoid layer is not used during training and only used at prediction?

Thank you

**Comment by wolny** [comment]
Hi there,

> So, when using dice, the final sigmoid layer is only used so that we can view the network output? Is that its only function?

No. Sigmoid normalization MUST be applied when training with `DiceLoss` or `GeneralizedDiceLoss`, but it's applied implicitly in the implementation (see: https://github.com/wolny/pytorch-3dunet/blob/master/unet3d/losses.py#L62). I used this convention for consistency with PyTorch, where e.g. `CrossEntropyLoss` also implicitly normalizes the input probabilities with softmax:

```
class CrossEntropyLoss(_WeightedLoss):
    r"""This criterion combines :func:`nn.LogSoftmax` and :func:`nn.NLLLoss` in one single class.
```
Using sigmoid/softmax explicitly in the model would incorrectly apply sigmoid/softmax twice: 1st during the forward pass through the model and 2nd during computation of the loss.
Since I want to be able to visualize normalized probabilities in tensorboard during training, I have to unfortunately invoke the the normalization layer explicitly.

> Would the network train similarly if the final sigmoid layer is not used during training and only used at prediction?
Also no, the training dynamics would be completely different. Skipping the sigmoid would mean that you're trying to fit logits (unnnormalized) to the targets, which would make it more difficult for the network. Training in such a way with CrossEntropy would probably converge somewhere, but it would be completely off for Dice, e.g. since logits are unnormalized there will be negative values in the network output, computing Dice score with negative inputs would incorrectly lower the loss.
